### ISSUE TITLE
Cannot start racket-repl: works only in racket-mode

### ISSUE THREAD

**Issue by kurniawano** [greghendershott/racket-mode#713]
Hi,
I'm new to racket and trying out racket-mode. I'm using Emacs 29.3.

Steps to reproduce: 
I installed racket-mode using package-install and set the following in my `init.el`.
```
(add-to-list 'auto-mode-alist '("\\.rkt\\'" . racket-mode))
(setq racket-racket-program "/Applications/Racket v8.13/bin/racket")
(setq racket-raco-program "/Applications/Racket v8.13/bin/racket")
```

I then open any `.rkt` file.

I can start `racket-mode` since I see the work "Racket" at my Emacs bar. However, every time I run `M-x racket-repl` it shows the following error.

```
racket-repl works only in racket-mode or racket-hash-lang-mode edit buffers
```

I'm not sure what's the issue. 

Below is the bug-report detail:

<details>
<h2>Package</h2>
<dl>
<dt>metadata</dt><dd><pre>(#s(package-desc racket-mode
                 (20240612 1256)
                 "Racket editing, REPL, and more"
                 ((emacs
                   (25 1)))
                 nil nil "/Users/oka_kurniawan/.emacs.d/elpa/racket-mode-20240612.1256"
                 ((:url . "https://www.racket-mode.com/")
                  (:maintainer "Greg Hendershott")
                  (:maintainers
                   ("Greg Hendershott"))
                  (:authors
                   ("Greg Hendershott" . "racket-mode-author@greghendershott.com"))
                  (:commit . "4cdbc1deb30c1625fadde762e55eb78c3acede53"))
                 nil))
</pre></dd>
<dt>package-archives</dt><dd><pre>(("gnu" . "https://elpa.gnu.org/packages/")
 ("nongnu" . "https://elpa.nongnu.org/nongnu/")
 ("tromey" . "http://tromey.com/elpa/")
 ("melpa" . "https://melpa.org/packages/"))
</pre></dd>
<dt>racket--el-source-dir</dt><dd><pre>"/Users/oka_kurniawan/.emacs.d/elpa/racket-mode-20240612.1256/"</pre></dd>
<dt>racket--rkt-source-dir</dt><dd><pre>"/Users/oka_kurniawan/.emacs.d/elpa/racket-mode-20240612.1256/racket/"</pre></dd>
</dl>
<h2>System values</h2>
<dl>
<dt>emacs-version</dt><dd><pre>"29.3"</pre></dd>
<dt>major-mode</dt><dd><pre>help-mode</pre></dd>
<dt>system-type</dt><dd><pre>darwin</pre></dd>
<dt>display-graphic-p</dt><dd><pre>t</pre></dd>
</dl>
<h2>Buffer values</h2>
<dl>
<dt>after-change-functions</dt><dd><pre>nil</pre></dd>
<dt>before-change-functions</dt><dd><pre>nil</pre></dd>
<dt>completion-at-point-functions</dt><dd><pre>(tags-completion-at-point-function)
</pre></dd>
<dt>eldoc-documentation-function</dt><dd><pre>eldoc-documentation-default</pre></dd>
<dt>font-lock-defaults</dt><dd><pre>nil</pre></dd>
<dt>pre-command-hook</dt><dd><pre>(tooltip-hide)
</pre></dd>
<dt>post-command-hook</dt><dd><pre>(global-font-lock-mode-check-buffers global-eldoc-mode-check-buffers global-display-line-numbers-mode-check-buffers ido-cr+-schedule-dynamic-collection-update global-hl-line-highlight mode-local-post-major-mode-change)
</pre></dd>
<dt>post-self-insert-hook</dt><dd><pre>(electric-indent-post-self-insert-function blink-paren-post-self-insert-function)
</pre></dd>
<dt>xref-backend-functions</dt><dd><pre>(etags--xref-backend)
</pre></dd>
</dl>
<h2>Racket Mode values</h2>
<dl>
<dt>racket--cmd-open-p</dt><dd><pre>t</pre></dd>
<dt>racket-after-run-hook</dt><dd><pre>nil</pre></dd>
<dt>racket-back-end-configurations</dt><dd><pre>((:directory "/" :racket-program nil :remote-source-dir nil :restart-watch-directories nil :windows nil))
</pre></dd>
<dt>racket-before-run-hook</dt><dd><pre>(racket-ansi-color-context-reset)
</pre></dd>
<dt>racket-browse-url-function</dt><dd><pre>racket-browse-url-using-temporary-file</pre></dd>
<dt>racket-command-timeout</dt><dd><pre>10</pre></dd>
<dt>racket-documentation-search-location</dt><dd><pre>"https://docs.racket-lang.org/search/index.html?q=%s"</pre></dd>
<dt>racket-error-context</dt><dd><pre>medium</pre></dd>
<dt>racket-expand-hiding</dt><dd><pre>standard</pre></dd>
<dt>racket-hash-lang-token-face-alist</dt><dd><pre>((constant . font-lock-constant-face)
 (error . error)
 (other . font-lock-doc-face)
 (keyword . font-lock-keyword-face)
 (hash-colon-keyword . racket-keyword-argument-face)
 (at . font-lock-doc-face))
</pre></dd>
<dt>racket-history-filter-regexp</dt><dd><pre>"\\`\\s *\\'"</pre></dd>
<dt>racket-imagemagick-props</dt><dd><pre>nil</pre></dd>
<dt>racket-images-do-not-use-svg</dt><dd><pre>nil</pre></dd>
<dt>racket-images-inline</dt><dd><pre>t</pre></dd>
<dt>racket-images-keep-last</dt><dd><pre>100</pre></dd>
<dt>racket-images-system-viewer</dt><dd><pre>"open"</pre></dd>
<dt>racket-indent-curly-as-sequence</dt><dd><pre>t</pre></dd>
<dt>racket-indent-sequence-depth</dt><dd><pre>0</pre></dd>
<dt>racket-logger-config</dt><dd><pre>((cm-accomplice . warning)
 (GC . info)
 (module-prefetch . warning)
 (optimizer . info)
 (racket/contract . error)
 (racket-mode-debugger . info)
 (sequence-specialization . info)
 (* . fatal))
</pre></dd>
<dt>racket-memory-limit</dt><dd><pre>2048</pre></dd>
<dt>racket-mode-hook</dt><dd><pre>nil</pre></dd>
<dt>racket-module-forms</dt><dd><pre>"\\s(\\(?:module[*+]?\\|library\\)"</pre></dd>
<dt>racket-pretty-lambda</dt><dd><pre>nil</pre></dd>
<dt>racket-pretty-print</dt><dd><pre>t</pre></dd>
<dt>racket-program</dt><dd><pre>"/Applications/Racket v8.13/bin/racket"</pre></dd>
<dt>racket-repl-buffer-name-function</dt><dd><pre>nil</pre></dd>
<dt>racket-repl-command-file</dt><dd><pre>"/Users/oka_kurniawan/.emacs.d/racket-mode/repl.rkt"</pre></dd>
<dt>racket-repl-history-directory</dt><dd><pre>"~/.emacs.d/racket-mode/"</pre></dd>
<dt>racket-repl-mode-hook</dt><dd><pre>nil</pre></dd>
<dt>racket-sexp-comment-fade</dt><dd><pre>0.5</pre></dd>
<dt>racket-shell-or-terminal-function</dt><dd><pre>racket-shell</pre></dd>
<dt>racket-show-functions</dt><dd><pre>(racket-show-pseudo-tooltip)
</pre></dd>
<dt>racket-smart-open-bracket-enable</dt><dd><pre>nil</pre></dd>
<dt>racket-submodules-to-run</dt><dd><pre>((test)
 (main))
</pre></dd>
<dt>racket-use-repl-submit-predicate</dt><dd><pre>nil</pre></dd>
<dt>racket-xp-add-binding-faces</dt><dd><pre>nil</pre></dd>
<dt>racket-xp-after-change-refresh-delay</dt><dd><pre>1</pre></dd>
<dt>racket-xp-highlight-unused-regexp</dt><dd><pre>"^[^_]"</pre></dd>
<dt>racket-xp-mode-lighter</dt><dd><pre>(:eval
 (racket--xp-mode-lighter))
</pre></dd>
</dl>
<h2>Minor modes</h2>
<dl>
<dt>enabled</dt><dd><pre>((auto-composition-mode)
 (auto-compression-mode)
 (auto-encryption-mode)
 (buffer-read-only)
 (display-line-numbers-mode)
 (editorconfig-mode)
 (file-name-shadow-mode)
 (font-lock-mode)
 (global-display-line-numbers-mode)
 (global-eldoc-mode)
 (global-font-lock-mode)
 (global-hl-line-mode)
 (ido-everywhere)
 (ido-ubiquitous-mode)
 (isearch-fold-quotes-mode)
 (line-number-mode)
 (mouse-wheel-mode)
 (projectile-mode)
 (recentf-mode)
 (save-place-mode)
 (semantic-minor-modes-format)
 (shell-dirtrack-mode)
 (show-paren-mode)
 (tool-bar-mode)
 (tooltip-mode)
 (transient-mark-mode))
</pre></dd>
<details><summary>Disabled minor modes</summary>
<dt>disabled</dt><dd><pre>((abbrev-mode)
 (archive-subfile-mode)
 (auto-fill-function)
 (auto-fill-mode)
 (auto-save-mode)
 (auto-save-visited-mode)
 (blink-cursor-mode)
 (buffer-face-mode)
 (button-mode)
 (cl-old-struct-compat-mode)
 (column-number-mode)
 (comint-fontify-input-mode)
 (compilation-minor-mode)
 (compilation-shell-minor-mode)
 (completion-in-region-mode)
 (context-menu-mode)
 (cursor-face-highlight-mode)
 (defining-kbd-macro)
 (diff-auto-refine-mode)
 (diff-minor-mode)
 (dired-hide-details-mode)
 (doc-view-minor-mode)
 (doc-view-presentation-mode)
 (eldoc-mode)
 (electric-indent-mode)
 (electric-layout-mode)
 (electric-pair-mode)
 (electric-quote-mode)
 (global-prettify-symbols-mode)
 (global-semantic-highlight-edits-mode)
 (global-semantic-highlight-func-mode)
 (global-semantic-show-parser-state-mode)
 (global-semantic-show-unmatched-syntax-mode)
 (global-semantic-stickyfunc-mode)
 (global-visual-line-mode)
 (gnus-dead-summary-mode)
 (gnus-undo-mode)
 (header-line-indent-mode)
 (hl-line-mode)
 (horizontal-scroll-bar-mode)
 (hs-minor-mode)
 (ibuffer-auto-mode)
 (ido-cr+-debug-mode)
 (image-minor-mode)
 (indent-tabs-mode)
 (isearch-mode)
 (jit-lock-debug-mode)
 (lock-file-mode)
 (lost-selection-mode)
 (mail-abbrevs-mode)
 (menu-bar-mode)
 (minibuffer-electric-default-mode)
 (mml-mode)
 (next-error-follow-minor-mode)
 (org-cdlatex-mode)
 (org-list-checkbox-radio-mode)
 (org-src-mode)
 (org-table-follow-field-mode)
 (org-table-header-line-mode)
 (orgtbl-mode)
 (outline-minor-mode)
 (overwrite-mode)
 (paragraph-indent-minor-mode)
 (paredit-mode)
 (prettify-symbols-mode)
 (racket-smart-open-bracket-mode)
 (racket-xp-mode)
 (read-extended-command-mode)
 (semantic-highlight-edits-mode)
 (semantic-highlight-func-mode)
 (semantic-mode)
 (semantic-show-parser-state-mode)
 (semantic-show-unmatched-syntax-mode)
 (semantic-stickyfunc-mode)
 (sh-electric-here-document-mode)
 (shell-highlight-undef-mode)
 (size-indication-mode)
 (tab-bar-history-mode)
 (tab-bar-mode)
 (tar-subfile-mode)
 (temp-buffer-resize-mode)
 (text-scale-mode)
 (treesit-explore-mode)
 (treesit-inspect-mode)
 (undelete-frame-mode)
 (url-handler-mode)
 (use-hard-newlines)
 (vc-dir-git-mode)
 (vc-parent-buffer)
 (view-mode)
 (visible-mode)
 (visual-line-mode)
 (window-divider-mode)
 (xref-etags-mode))
</pre></dd>
</details>
</dl>
</details>

**Comment by greghendershott** [comment]
You have a buffer for a `.rkt` file, and `racket-mode` is automatically enabled in that buffer; you can see `Racket` in the mode line for the buffer. Great!

Next, if you want to run that `.rkt` program, try using `M-x racket-run`. That should open (or reuse) a `racket-repl-mode` buffer, and run the program in it.

- Does that work when you try it?
- Is this actually what you wanted to accomplish?

---

Using `M-x racket-repl` isn't intended to run a program. Instead it's a way to get a kind of "blank" REPL just to noodle around in. See https://www.racket-mode.com/#racket_002drepl.

`M-x racket-repl` should work from _any_ buffer -- without giving you the error message that it works only in `racket-mode` or `racket-hash-lang-mode` buffers. I can't reproduce that; I don't get that error message. I'm trying to imagine how that could occur in your case... no theories quite yet.

---

TL;DR: I think maybe you just want to use `racket-run` (instead of `racket-repl`) and that will solve your "real", actual problem? OTOH what you report about `racket-repl` sounds like a bug; I just can't reproduce it quite yet.

**Comment by greghendershott** [comment]
One tiny point. I'm 99% it's unrelated to the problem you're reporting. Just FYI:

> I installed racket-mode using package-install and set the following in my `init.el`.
> 
> ```
> (add-to-list 'auto-mode-alist '("\\.rkt\\'" . racket-mode))
> (setq racket-racket-program "/Applications/Racket v8.13/bin/racket")
> (setq racket-raco-program "/Applications/Racket v8.13/bin/racket")
> ```

You just need:

```
(add-to-list 'auto-mode-alist '("\\.rkt\\'" . racket-mode))
(setq racket-program "/Applications/Racket v8.13/bin/racket")
```

The `racket-racket-program` and `racket-raco-program` variables have been obsolete since 2017. Just `racket-program` suffices for the `racket` executable. See https://www.racket-mode.com/#Configure.

**Comment by kurniawano** [comment]
@greghendershott . Thanks for the tips. I removed the line on racket-raco-program. 
I can run M-x racket-run and the REPL appears. Yay.

However, M-x racket-repl still does not open blank REPL. 
This is what I did. 
* I open rkt file.
* make sure it's in racket-mode.
* Run M-x racket-run
* I can see the repl
* I kill REPL using M-x kill-buffer
* Run M-x racket-repl
* It gives the same error "racket-repl works only in racket-mode or racket-hash-lang-mode edit buffers"

I'm not sure why. I think for run running "racket-run" works good for me. But I find it strange. I didn't encounter this another Mac machine which runs on Emacs 28. It works fine there to open blank REPL.

**Comment by greghendershott** [comment]
Thanks for the information!

As its doc string explains...

> Show a Racket REPL buffer in some window.
> 
> The intended use of Racket Mode's REPL is that you find-file
> some specific file, then run it using a command like racket-run
> or racket-run-module-at-point. The resulting REPL will
> correspond to those definitions and match your expectations.
> 
> Therefore this racket-repl command -- which is intended as a
> convenience for people who want to "just get a quick scratch
> REPL" -- is actually implemented as running the file named in
> the customization variable racket-repl-command-file. When that
> file doesn't exist, it is created to contain just "#lang
> racket/base". You may edit the file to use a different lang,
> require other modules, or whatever.

... `racket-repl` works by running a .rkt file -- the one named by the customization variable `racket-repl-command-file`.

This defaults to `~/.emacs.d/racket-mode/repl.rkt`, and is automatically created the first time if it doesn't exist.

**Question**: I wonder what happens if you try to open this file directly with `M-x find-file` -- does the buffer use `racket-mode` (do you see "Racket" down in the buffer's mode line)?

---

> I didn't encounter this another Mac machine which runs on Emacs 28. It works fine there to open blank REPL.

Interesting. Maybe something changed after Emacs 28. I use 28.2 as my daily driver, which is what I tried.

I'll rebuild Emacs from source, which is version 30-something these days, and see if I can reproduce the problem, there.

**Comment by kurniawano** [comment]
@greghendershott , I can open `repl.rkt` and it contains the following.
```
;; Used by M-x racket-repl; you may edit
#lang racket/base
```

Somehow it didn't use `racket-mode` when I open the file but I can activate it after that using `M-x racket-mode`. Now I recall that sometimes when I open other .rkt file it didn't use `racket-mode` and I basically just manually activate it.

**Comment by greghendershott** [comment]
TL;DR:

1. I have a commit a56a79d (not yet merged) that forces the `racket-repl-command-file` buffer's major mode to `racket-mode`, which should avoid any error message even if someone's configuration doesn't reliably/consistently do so for `.rkt` file buffers. I think that closes the reported issue, per se.

2. At the same time, I'm happy to keep working with you to try to figure out what's going on with your configuration -- determine why it's not already doing this consistently/reliably. If you'd like to work on this, please start by answering some of my questions from my previous post, and we can take it from there!

**Comment by greghendershott** [comment]
I went ahead and merged the commit.

---

Someday if you have time, I'd be happy to help try to troubleshoot your configuration. 

- Because of the inconsistency -- `racket-mode` is only _sometimes_ chosen for `.rkt` files -- my wild guess is you have `auto-mode-list` in a `.dir-locals.el` file somewhere, that picks some other mode like idk `scheme-mode`?

- I suppose it could be one of the other vars like `major-mode-map-alist` (although IIUC that should _consistently_ not choose racket-mode)?

- I suppose it could even be some regexp in `magic-mode-alist` or another var, that is subtly wrong, and only matches _some_ `.rkt` filenames??

But if you're tired of this whole issue and want to move on, I'd understand that, too.

**Comment by kurniawano** [comment]
@greghendershott , sorry I was away for a few days. You are right. I wrongly set the paredit as major mode and it overwrite it. I have changed it to minor mode and now it always open .rkt files in racket-mode. I can run racket-repl also now. Thanks so much.

**Comment by greghendershott** [comment]
> I wrongly set the paredit as major mode and it overwrite it. I have changed it to minor mode and now it always open .rkt files in racket-mode.

Oh I see, that makes sense.

Glad it's working for you now.

For me it was at least a chance to make `racket-repl` a tiny bit more "resilient", as well as dig into some Emacs docs and source code learning a bit more about exactly how major modes get picked.
